### PR TITLE
Feature/csl1070 peer discovery prettify

### PR DESCRIPTION
**Pull request by volhovM**
This adds the following features:
* Discovery mechanism is abstracted in `infra/Pos/Discovery/Class` as `MonadDiscovery`.
* Two instances of `MonadDiscovery` are added in form of Reader transformers -- see `infra/Pos/Discovery/Holders`. One of them is KademliaDHT (holds kademlia instance inside), the other is static peers reader.
* `m (Set NodeId)` parameter is removed from every argument passing place where it was.
* `Pos.WorkMode` has changed -- now `ProductionMode` contains kademlia transformer on the top, so as `StatsMode`. `StaticMode` is introduced with is prod mode with static peers instead of dht.
* Some leftovers are cleared that were not done in 995 PR (renamings, moving modules around, styling).
* `Pos.Launcher.Resources` is removed as it contained transport plus discovery functionality only. All previous usages now have `Transport` inlined (instead of `RealModeResources`).
* Wallet part and executables are changed respectively (but only cosmetic improvements).

### CODE REVIEW

File: infra/Pos/Communication/Relay/Logic.hs [hunk 1]
```diff
@@ -314,16 +315,18 @@ addToRelayQueue inv = do
...
-             ( Mockable Throw m
-             , RelayWorkMode m
-             , MonadMask m
-             , MonadReportingMem m
-             )
-             => m (Set NodeId) -> OutSpecs -> ([WorkerSpec m], OutSpecs)
-relayWorkers getPeers allOutSpecs =
+relayWorkers
+    :: forall m.
+       ( Mockable Throw m
```
Comment (int-index) [inline infra/Pos/Communication/Relay/Logic.hs]: Ah, the rare specimen, 7-space indent.

**Comment by avieth** [comment]
> Pos.Launcher.WorkMode has changed -- now ProductionMode contains kademlia transformer on the top, so as StatsMode. StaticMode is introduced with is prod mode with static peers instead of dht.

Er, does this not completely undo the work that I did to abstract discovery?

File: infra/Pos/Communication/Relay/Logic.hs (hunk 1 above)
Comment (volhovM) [inline infra/Pos/Communication/Relay/Logic.hs]: Actually it's default/hindent styling. `::` and following `=>` and `->`s are indented to 4 spaces what's important.

**Comment by gromakovsky** [comment]
> `Pos.Launcher.WorkMode`

It was surprising, but looks like it's a typo.

File: infra/Pos/Discovery/Class.hs [hunk 2]
```diff
@@ -0,0 +1,20 @@
+{-# LANGUAGE RankNTypes #-}
+
+-- | Abstraction over possibility to retrieve neighbors.
+
+module Pos.Discovery.Class
+       ( MonadDiscovery (..)
+       ) where
+
+import           Control.Monad.Trans              (MonadTrans)
+import           Pos.Communication.Types.Protocol (NodeId)
```
Comment (gromakovsky) [inline infra/Pos/Discovery/Class.hs]: Wrong grouping btw, but I don't care much.

File: infra/Pos/Discovery/Holders.hs [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+       , runDiscoveryKademliaT
+       ) where
+
+import qualified Control.Monad.Ether              as Ether
+import qualified Control.Monad.Ether              as Ether.E
+import qualified Data.Set                         as S (fromList)
+import           Mockable                         (Async, Catch, Mockables, Promise,
+                                                   Throw)
+import           System.Wlog                      (WithLogger)
+import           Universum
```
Comment (gromakovsky) [inline infra/Pos/Discovery/Holders.hs]: But anyway, why don't you put `Universum` as the first import? I don't insist on fixing it now, but we agreed to do it long ago :/

File: infra/Pos/Discovery/Class.hs [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+import           Universum
+
+class (Monad m) => MonadDiscovery m where
+    getPeers  :: m (Set NodeId)
+    findPeers :: m (Set NodeId)
```
Comment (gromakovsky) [inline infra/Pos/Discovery/Class.hs]: Can you add documentation of these functions? It's not obvious what `findPeers` should return, for example.

File: infra/Pos/Discovery/Holders.hs [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+----------------------------------------------------------------------------
+-- Constant peers
+----------------------------------------------------------------------------
+
+type DiscoveryConstT m = Ether.E.ReaderT DiscoveryTag (Set NodeId) m
+
+instance (Monad m) => MonadDiscovery (DiscoveryConstT m) where
+    getPeers = Ether.E.ask (Proxy @DiscoveryTag)
+    findPeers = pure mempty
```
Comment (gromakovsky) [inline infra/Pos/Discovery/Holders.hs]: Is it correct? `findPeers` for `DiscoveryKademliaT` returns result of `getPeers`, but this one returns `mempty`. That's why I am asking for documentation.

File: src/Pos/Ssc/NistBeacon.hs [hunk 6]
```diff
@@ -45,31 +46,30 @@ instance Ssc SscNistBeacon where
...
 instance SscListenersClass SscNistBeacon where
     sscListeners = return $ Tagged ([], mempty)
     sscStubListeners = Tagged ([], mempty)
 
 instance SscLocalDataClass SscNistBeacon where
-    sscGetLocalPayloadQ _ = pure ()
+    sscGetLocalPayloadQ _ = pass
     sscNormalizeU _ _ _ = pass
-    sscNewLocalData = pure ()
+    sscNewLocalData = pass
```
Comment (gromakovsky) [inline src/Pos/Ssc/NistBeacon.hs]: I don't like some changes here. `sscNewLocalData` returns local data which is defined as `()`.
`pass` has a semantics of `do nothing`.
`pure ()` has a semantics of `return ()`.

It's a clumsy explanation, but I think you got it.

Comment (gromakovsky) [inline src/Pos/Ssc/NistBeacon.hs]: Obviously `pass = pure ()`, but I treat them differently.

File: src/node/Main.hs [hunk 7]
```diff
@@ -130,54 +132,56 @@ action kademliaInst args@Args {..} res = do
...
     putText $ "If stats is on: " <> show enableStats
+    let wDhtWorkersStats :: ([WorkerSpec (StatsMode ssc')], OutSpecs)
+        wDhtWorkersStats =
+            first (map $ wrapActionSpec $ "worker" <> "dht") (dhtWorkers kademliaInst)
+    let wDhtWorkersProd :: ([WorkerSpec (ProductionMode ssc')], OutSpecs)
+        wDhtWorkersProd =
+            first (map $ wrapActionSpec $ "worker" <> "dht") (dhtWorkers kademliaInst)
     case (enableStats, CLI.sscAlgo commonArgs) of
-               
+        -- This is a terrible code duplication. Abstract it when possible.
```
Comment (gromakovsky) [inline src/node/Main.hs]: Affirmative.

File: src/Pos/Launcher/Runner.hs [hunk 8]
```diff
@@ -386,38 +444,29 @@ bracketTransport
...
-    -> (RealModeResources Production -> Production a)
+    -> (Transport Production -> Production a)
     -> IO a
-bracketResources bp tcpAddr staticPeers action =
+bracketResources bp tcpAddr action =
     loggerBracket (bpLoggingParams bp) .
     runProduction $
         -- Both the DHT and Transport are problematic here:
         -- 1. We assume you'll want a DHT.
         -- 2. We assume your transport takes an IP/port.
```
Comment (volhovM) [inline src/Pos/Launcher/Runner.hs]: @avieth can you please comment on what did you mean? I see it was introduced in d8618bf9e48f8ade96de0acf069bdf2c2a30e4ba.

**Pull request merged**